polling_stations/apps/data_finder/helpers/directions.py

```diff
@@ -97,12 +110,69 @@ def get_route(self, start, end):
         )
 
 
+class MapboxDirectionsClient(DirectionsClient):
+    precision = 5
+
+    def get_data(self, url):
+        resp = requests.get(url)
+        if resp.status_code != 200:
+            raise DirectionsException(
+                "Mapbox Directions API error: HTTP status code %i" % resp.status_code
+            )
+        return resp.json()
+
+    def get_route(self, start, end):
+        distance_km = get_distance(start, end)
+        if distance_km > 1.5:
+            transport_verb = {"base": "drive", "gerund": "driving-traffic"}
+        else:
+            transport_verb = {"base": "walk", "gerund": "walking"}
+
+        url = "{base_url}/{profile}/{coordinates}?&alternatives=false&geometries=polyline&steps=false&access_token={key}".format(
+            base_url=settings.BASE_MAPBOX_URL,
+            profile=transport_verb["gerund"],
+            coordinates="{start_lon},{start_lat};{end_lon},{end_lat}".format(
+                start_lon=start.x, start_lat=start.y, end_lon=end.x, end_lat=end.y
+            ),
+            key=settings.MAPBOX_API_KEY,
+        )
+
+        directions = self.get_data(url)
+
+        if directions["code"] != "Ok":
+            raise DirectionsException(
+                "Mapbox Directions API error: {}".format(directions["code"])
+            )
+
+        route = directions["routes"][0]["geometry"]
+
+        time = append_minute(math.ceil(directions["routes"][0]["duration"] / 60))
+
+        dist = append_miles(
+            math.ceil(directions["routes"][0]["distance"] * 0.0006213712)
+        )
+
+        return Directions(
+            time,
+            dist,
+            transport_verb["base"],
+            json.dumps(route),
+            self.precision,
+            "Mapbox",
+        )
+
+
 class DirectionsHelper:
     def get_directions(self, **kwargs):
         if kwargs["start_location"] and kwargs["end_location"]:
-            clients = (GoogleDirectionsClient(),)
+
+            clients = (
+                MapboxDirectionsClient(),
+                GoogleDirectionsClient(),
+            )
```

Comment:
Workaround idea I was thinking about on standup - if you had a conditional here
```
if settings.SKIP_MAPBOX:
    clients = (GoogleDirectionsClient(),)
```
But would still like to understand how to properly mock within the integration tests

Reply:
In general I think moving to factories etc for test data is a good move for unit tests :+1:  VCR is useful in certain limited situations - there's a bit more discussion about where VCR works well/badly (IMHO) in https://github.com/DemocracyClub/aggregator-api/pull/56#issuecomment-507747159

That said, with the end-to-end tests, they are a blackbox test: Because you are literally spinning up an application server and a (headless) web browser and making HTTP calls to the app server, it is much harder to "mock" individual components inside the server so easily like you can in a unit test so VCR (or something similar) is the easy option for mocking external interactions. Its a blackbox test and a blackbox approach to mocking. I'd say your options are basically:

- Create VCR responses for the mapbox calls (is there a serious downside to this you are trying to avoid?)

- If you just want to call gmaps, use the test settings to configure what services you are going to call. A `SKIP_MAPBOX` setting is fairly crude, but settings are just python, so you could literally set `DIRECTIONS_CLIENTS = (MapboxDirectionsClient(), GoogleDirectionsClient(),)` in `base.py` and `DIRECTIONS_CLIENTS = (GoogleDirectionsClient(),)` in `testing.py` then set `clients = settings.DIRECTIONS_CLIENTS` here. Another approach would be to determine which clients you're going to call based on which services you have keys for e.g:

```py
clients = []

if settings.MAPBOX_API_KEY:
    clients.append(MapboxDirectionsClient())

# you can actually bin the ability to rotate multiple google keys now - we only need one
if settings.GOOGLE_API_KEYS:
    clients.append(GoogleDirectionsClient())
```

..and then in `testing.py` you can set a dummy key for whichever service you want to use/mock under test, That said, for dev purposes it might be useful to retain the ability to do low scale anonymous usage on some service without having a key for either.

In fact, you might be able to just skip this interaction completely when running the end-to-end tests (I can't remember if we actually make any assertions about map/directions in the end-to-end tests?). If not you could give neither a key or set `DIRECTIONS_CLIENTS = []` under test and bin the mocks for a directions service.

Anyway, there's a couple of other ideas for controlling it with test settings..